### DISCUSSION TITLE
report level

### DISCUSSION THREAD

**Discussion by samuelhurtado** [EdNekebno/DGTCentaurMods#149] [Ideas]
Hi. Very interesting project, thanks for sharing everything here.

I bought three DGT Centaur boards: one for me, one for my father, one for my nephew. We all love the board, and we all have the same problem with it: lack of feedback about our level improvements. Being able to install a new engine with finer levels could sort this out, but for me a simpler change would be better: just have the board report, at the end of the game, a rating telling us how well it played. The adaptative code must have a variable somewhere choosing that, no idea why they wouldn't report it at the end.

My question is: in the DGT python files, is that variable accessible somewhere, or is it inside the encrypted parts of the code? Because I'd rather make a small change than install a hardware mod and fully different software.

Thanks again.

**Comment by wormstein** [comment]
Hi,
sorry to dissapoint you but I don't think there is such a parameter. And I will tell you why.
Original centaur software stores all the settings inside a file in `/settings`. This is the content of it:
```
{"LargeFont": false, "Rotated": false, "Language": "EN", "LightLevel": 3, "PlaySound": false}
```
Nothing to suggest such a parameter. And it also saves the state of the game inside other data files between powercycles. The python code is a compiled version that gets extracted on each boot and executed. Saying that, I don't think they can store that variable inside the code to survive between reboots...

Adaptive code is just a marketing "strategy" :)

We don't have access to the original python code.

Cheers

**Comment by samuelhurtado** [comment]
Thanks. That means it must be inside the encrypted code then: there must be a "play the Nth best move" variable somewhere, but if we can't access it, it's as if it didn't exist.

I may have to try the full hardware+software mod in one of the boards (mine: I haven't used it in months because I get better motivation playing against people online than OTB against a machine that doesn't tell me how I'm doing.

**Comment by wormstein** [comment]
The code will execute the exact same way on each boot. Because it's compiled. There is no way to store a persistent veriable between reboots inside the code. That's way the board keeps it's settings and game state outside on files.
The adaptive mode likely happens on consecutive matches between powercycles. On a new boot, it will start on the level you selected in the settings menu, then it will start adapting with each game.

-A.

**Comment by samuelhurtado** [comment]
I received email notification of a very interesting comment but I don't see it here, I guess the author deleted it, maybe because it had guesswork. But it had a very interesting part: "the original DGT Centaur code asks stockfish for possible moves with the scores. Rather than using the best possible move, it will pick a different move that moves the score closer to 0"

From playing with it, I agree that this is probably what it's doing. If I could access that part of the code, I'd report, at the end of the game, something like "on average the board played the 7.3th-best move (range: 2nd to 14th)". But if it's inside the encrypted part of the code, we can't access it. I'll try the full mod... once raspberry boards are back in stock: they see to be unavailable everywhere right now

**Comment by wormstein** [comment]
I think it's simpler than that, The engine is not reporting several possible moves to pick one because this will require much more computing power. Each calculated possible move at a different ELO level requires a full re-evaluation of the position.... clearly this is not the case based on how fast the board responds.

How I think this works is the engine scores your move then report it back. The adaptive mode will use the score when it asks the engine for the his next move. This is possible by capping the ELO level of the engine.

I would rather report that score at the end of the match but there is no way to access their code.

This sore is kept as a variable as long as the board is on, even between matches. When you turn it off the score will reset to whatever level you have selected in settings.

-A

**Comment by samuelhurtado** [comment]
Whatever the details, what's clear is that it happens in the encrypted part, so only they can add the end-of-game level report. And they won't do it (I wrote asking for that feature, many months ago).

Thanks again.

**Comment by wormstein** [comment]
Maybe in future versions of the board, not on the current ones because Pi inside has no connectivity to do an OTA update. And the miniUSB connector goes to the controller, not the Pi.
